dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs

```diff
@@ -94,255 +95,454 @@ public Plan CreatePlan(string goal)
         return plan;
     }
 
+    /// <summary>
+    /// Execute a plan
+    /// </summary>
+    /// <param name="question">The question to answer</param>
+    /// <param name="context">The context to use</param>
+    /// <param name="token">The cancellation token</param>
+    /// <returns>The context with the result</returns>
+    /// <exception cref="SKException">No AIService available for getting completions.</exception>
     [SKFunction, SKName("ExecutePlan"), Description("Execute a plan")]
     public async Task<SKContext> ExecutePlanAsync(
         [Description("The question to answer")]
         string question,
-        [Description("List of tool descriptions")]
-        string functionDescriptions,
-        SKContext context)
+        SKContext context,
+        CancellationToken token = default)
     {
+        if (string.IsNullOrEmpty(question))
+        {
+            context.Variables.Update("Question not found.");
+            return context;
+        }
+
+        ChatHistory chatHistory = await this.InitializeChatHistoryAsync(this.CreateChatHistory(this._kernel, out var aiService), aiService, context).ConfigureAwait(false);
+
+        if (aiService is null)
+        {
+            throw new SKException("No AIService available for getting completions.");
+        }
+
+        if (chatHistory is null)
+        {
+            throw new SKException("ChatHistory is null.");
+        }
+
+        var startingMessageCount = chatHistory.Messages.Count;
+
         var stepsTaken = new List<SystemStep>();
-        if (!string.IsNullOrEmpty(question))
+        SystemStep? lastStep = null;
+
+        var GetNextStepAsync = async () =>
```

Comment:
I'm wondering if it would be possible to keep all these delegates as separate methods to simplify root `ExecutePlanAsync` method.

Reply:
Agreed, started to go down this path. I think the right thing to do will be creating a Parser and Engine to simplify responsibilities.